### PR TITLE
feat: read accumulative result files

### PR DESCRIPTION
**Pull request by andrewferlitsch**
**REQUIRED:** Add a summary of your PR here, typically including why the change is needed and what was changed. Include any design alternatives for discussion purposes.

<br>
Work for followup PR:

- Replace JSON intermediate storage with serialized/deserialized NetworkExecutionResult.
- Fine-tune the aging algorithm for selecting notebooks.
- Add time window in aging algorithm.
- Read in bulk build_results. **DONE**
<br><br><br>

**REQUIRED:** Fill out the below checklists or remove if irrelevant
1. If you are opening a PR for `Official Notebooks` under the [notebooks/official](https://github.com/GoogleCloudPlatform/vertex-ai-samples/tree/main/notebooks/official) folder, follow this mandatory checklist:
- [ ] Use the [notebook template](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/notebook_template.ipynb) as a starting point.
- [ ] Follow the style and grammar rules outlined in the above notebook template.
- [ ] Verify the notebook runs successfully in Colab since the automated tests cannot guarantee this even when it passes.
- [ ] Passes all the required automated checks. You can locally test for formatting and linting with these [instructions](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/CONTRIBUTING.md#code-quality-checks).
- [ ] You have consulted with a tech writer to see if tech writer review is necessary. If so, the notebook has been reviewed by a tech writer, and they have approved it.
- [ ] This notebook has been added to the [CODEOWNERS](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/official/CODEOWNERS) file under the `Official Notebooks` section, pointing to the author or the author's team.
- [ ] The Jupyter notebook cleans up any artifacts it has created (datasets, ML models, endpoints, etc) so as not to eat up unnecessary resources.

<br>

2. If you are opening a PR for `Community Notebooks` under the [notebooks/community](https://github.com/GoogleCloudPlatform/vertex-ai-samples/tree/main/notebooks/community) folder:
- [ ] This notebook has been added to the [CODEOWNERS](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/community/CODEOWNERS) file under the `Community Notebooks` section, pointing to the author or the author's team.
- [ ] Passes all the required formatting and linting checks. You can locally test with these [instructions](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/CONTRIBUTING.md#code-quality-checks).

<br>

3. If you are opening a PR for `Community Content` under the [community-content](https://github.com/GoogleCloudPlatform/vertex-ai-samples/tree/main/community-content) folder:
- [ ] Make sure your main `Content Directory Name` is descriptive, informative, and includes some of the key products and attributes of your content, so that it is differentiable from other content
- [ ] The main content directory has been added to the [CODEOWNERS](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/community-content/CODEOWNERS) file under the `Community Content` section, pointing to the author or the author's team.
- [ ] Passes all the required formatting and linting checks. You can locally test with these [instructions](https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/CONTRIBUTING.md#code-quality-checks).

### CODE REVIEW

File: .cloud-build/execute_changed_notebooks_cli.py [hunk 1]
```diff
@@ -137,7 +137,11 @@ def str2bool(v):
 
 
 results_bucket = f"{args.artifacts_bucket}"
-results_file = f"{args.build_id}.json"
+# artifacts_bucket may get set by trigger to a full gs:// folder path
+if results_bucket.startswith("gs://"):
+    results_bucket = results_bucket[5:]
+results_bucket = results_bucket.split('/')[0]
+results_file = f"build_results/{args.build_id}.json"
```
Comment (ivanmkc) [inline .cloud-build/execute_changed_notebooks_cli.py:144]: Good

File: .cloud-build/execute_changed_notebooks_helper.py [hunk 2]
```diff
@@ -95,8 +95,14 @@ def load_results(results_bucket: str,
...
-        content = util.download_blob_into_memory(results_bucket, results_file, download_as_text=True)
-        accumulative_results = json.loads(content)
+        client = storage.Client()
+        bucket = client.bucket(results_bucket)
+
+        build_results_dir = os.path.dirname(results_file)
+        blobs = client.list_blobs(results_bucket, prefix=build_results_dir)
+        for blob in blobs:
+            content = util.download_blob_into_memory(results_bucket, blob.name, download_as_text=True)
+            accumulative_results = {**accumulative_results, **json.loads(content)}
```
Comment (ivanmkc) [inline .cloud-build/execute_changed_notebooks_helper.py:105]: Strange but I guess you'll be replacing with a list of `NotebookExecutionResults`?

I'm not understanding what this line does.

File: .cloud-build/execute_changed_notebooks_helper.py [hunk 3]
```diff
@@ -118,7 +124,15 @@ def select_notebook(changed_notebook: str,
...
-    return (random.randint(1, 100) * (1 + (fail_count / (pass_count + fail_count))) < test_percent)
+    inferred_failure_rate = fail_count / (pass_count + fail_count)
+
+    # If failure rate is high, the chance of testing should be higher
+    should_test_due_to_failure = random.uniform(0, 1) < inferred_failure_rate
+
+    # Additionally, only test a percentage of these
+    should_test_due_to_random_subset = random.uniform(0, 1) < test_percent
+
+    return should_test_due_to_failure or should_test_due_to_random_subset
```
Comment (ivanmkc) [inline .cloud-build/execute_changed_notebooks_helper.py:135]: `should_test_due_to_failure and should_test_due_to_random_subset` makes more sense if you only want to test a subset of the testable notebooks.

File: .cloud-build/execute_changed_notebooks_helper.py [hunk 4]
```diff
@@ -379,39 +393,30 @@ def get_changed_notebooks(
...
-            accumulative_results[result.path] = {
-                    'duration': result.duration.total_seconds(),
-                    'start_time': str(result.start_time),
-                    'passed': pass_count,
-                    'failed': fail_count
-                    }
-            print(f"adding {result.path}")
+            pass_count = 0
+            fail_count = 1
+        build_results[result.path] = {
```
Comment (ivanmkc) [inline .cloud-build/execute_changed_notebooks_helper.py:410]: The accumulation should only be done on the read path, i.e. `load_results` not in `_save_results`

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 2 above)
Comment (andrewferlitsch) [reply .cloud-build/execute_changed_notebooks_helper.py:105]: Each build outputs its own results file (ie., {build_id}.json) into the same GCS directory (../build_results). I scan the directory, load each JSON result file and then merge it with accumulative results.

For example, let's say notebook foo.ipynb appears in three {build_id}.json files, with 2 passes and 1 fail. Then accumulative results would look like:

{'foobar.ipynb': {'passed': 2, 'failed': 1, .... }

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 3 above)
Comment (andrewferlitsch) [reply .cloud-build/execute_changed_notebooks_helper.py:135]: This code is only reachable when --test_percent < 100

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 4 above)
Comment (andrewferlitsch) [reply .cloud-build/execute_changed_notebooks_helper.py:410]: Its not accumulating. The loop is converting the NotebookExecutionResult objects into the intermediate JSON format I am saving on disk.

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 3 above)
Comment (ivanmkc) [reply .cloud-build/execute_changed_notebooks_helper.py:135]: My comment is still relevant.

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 4 above)
Comment (ivanmkc) [reply .cloud-build/execute_changed_notebooks_helper.py:410]: You are accumulating pass count and failure count. That should be dinner done on the read path.

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 3 above)
Comment (andrewferlitsch) [reply .cloud-build/execute_changed_notebooks_helper.py:135]: I think that will require another --argument

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 4 above)
Comment (andrewferlitsch) [reply .cloud-build/execute_changed_notebooks_helper.py:410]: looked at the code again. It is only setting pass/fail to 0 or 1 (based on result.is_pass). I don't see any accumulating. Should only record results for this build.

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 3 above)
Comment (andrewferlitsch) [reply .cloud-build/execute_changed_notebooks_helper.py:135]: updated

Comment (ivanmkc) [reply .cloud-build/execute_changed_notebooks_helper.py:135]: No, that’s not required.
I’m just asking you to change ‘or’ to ‘and’.

We should strive to reduce complexity/parameters where possible.

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 4 above)
Comment (ivanmkc) [reply .cloud-build/execute_changed_notebooks_helper.py:410]: Okay great, then use a bool and rename to is_pass since we don’t need an int.

**Review by ivanmkc — COMMENTED** [review]
Unit tests should accompany functions with no side effects like select_notebook, since you don’t need mocks. 

Since this method is probabilistic though, I think you’ll need to do Monte Carlo simulations to properly test it… let’s discuss

**Comment by ivanmkc** [comment]
We should at least test the boundary conditions such as failure rate being 100% or 0% and same for test percentage.

File: .cloud-build/execute_changed_notebooks_cli.py [hunk 5]
```diff
@@ -47,6 +47,12 @@ def str2bool(v):
     required=False,
     default=100,
 )
+parser.add_argument(
```
Comment (ivanmkc) [inline .cloud-build/execute_changed_notebooks_cli.py:50]: Can remove. See other comment.

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 3 above)
Comment (andrewferlitsch) [reply .cloud-build/execute_changed_notebooks_helper.py:135]: done

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 4 above)
Comment (andrewferlitsch) [reply .cloud-build/execute_changed_notebooks_helper.py:410]: Yeah, that won't work. If I save as bool in each build results, then that breaks the code in load_results() where I accumulate the counts across plural build results. Can accumulate an int, but not a bool.

Comment (ivanmkc) [reply .cloud-build/execute_changed_notebooks_helper.py:416]: Sure okay, let's give it a try. Thanks for hearing me out.

Comment (ivanmkc) [reply .cloud-build/execute_changed_notebooks_helper.py:416]: It would work if you modify the read path to just count the number of True's.
The problem with the current approach is that the `pass_count` and `fail_count` is always 1, so there's no need for an int, which increases complexity (e.g. what happens another engineer comes along and changes the write path behavour, will the read-path still work?).

File: .cloud-build/execute_changed_notebooks_cli.py [hunk 6]
```diff
@@ -137,13 +137,20 @@ def str2bool(v):
...
+    results_bucket = results_bucket[5:]
+results_bucket = results_bucket.split('/')[0]
+results_file = f"build_results/{args.build_id}.json"
 
 if args.test_percent == 100:
     notebooks = changed_notebooks
     accumulative_results = {}
 else:
-    accumulative_results = execute_changed_notebooks_helper.load_results(results_bucket, results_file)
+    if args.test_filter:
```
Comment (ivanmkc) [inline .cloud-build/execute_changed_notebooks_cli.py:150]: This arg is removed, so this is unneeded.

File: .cloud-build/execute_changed_notebooks_helper.py (hunk 4 above)
Comment (ivanmkc) [reply .cloud-build/execute_changed_notebooks_helper.py:416]: I won't nitpick on this incessantly though so will defer to you.

**Review by ivanmkc — APPROVED** [review]
LGTM
